internal/terraform/node_action.go

```diff
@@ -0,0 +1,272 @@
+		if !allowActions {
+			summary := fmt.Sprintf("Actions experiment not enabled for module %s", module)
+			if module.IsRoot() {
+				summary = "Actions experiment not enabled"
+			}
+			return nil, diags.Append(tfdiags.Sourceless(
+				tfdiags.Error,
+				summary,
+				"The actions experiment must be enabled in order to use actions in your configuration. You can enable it by adding a terraform { experiments = [actions] } block to your configuration.",
+			))
+		}
+
+		// The rest of our work here needs to know which module instance it's
+		// working in, so that it can evaluate expressions in the appropriate scope.
+		_, err := n.expandActionInstances(ctx, absActAddr, &g)
+		diags = diags.Append(err)
+	}
+
+	return &g, diags
+}
+
+func (n *nodeExpandActionDeclaration) expandActionInstances(globalCtx EvalContext, actAddr addrs.AbsAction, g *Graph) ([]addrs.AbsActionInstance, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	// The rest of our work here needs to know which module instance it's
+	// working in, so that it can evaluate expressions in the appropriate scope.
+	moduleCtx := evalContextForModuleInstance(globalCtx, actAddr.Module)
+
+	// writeResourceState is responsible for informing the expander of what
+	// repetition mode this resource has, which allows expander.ExpandResource
+	// to work below.
+	moreDiags := n.recordActionData(moduleCtx, actAddr)
+	diags = diags.Append(moreDiags)
+	if moreDiags.HasErrors() {
+		return nil, diags
+	}
+
+	// Before we expand our resource into potentially many resource instances,
+	// we'll verify that any mention of this resource in n.forceReplace is
+	// consistent with the repetition mode of the resource. In other words,
+	// we're aiming to catch a situation where naming a particular resource
+	// instance would require an instance key but the given address has none.
+	expander := moduleCtx.InstanceExpander()
+	instanceAddrs := expander.ExpandAction(actAddr)
+
+	// NOTE: The actual interpretation of n.forceReplace to produce replace
+	// actions is in the per-instance function we're about to call, because
+	// we need to evaluate it on a per-instance basis.
+
+	// Our graph builder mechanism expects to always be constructing new
+	// graphs rather than adding to existing ones, so we'll first
+	// construct a subgraph just for this individual modules's instances and
+	// then we'll steal all of its nodes and edges to incorporate into our
+	// main graph which contains all of the resource instances together.
+	instG, instDiags := n.actionInstanceSubgraph(actAddr, instanceAddrs)
+	if instDiags.HasErrors() {
+		diags = diags.Append(instDiags)
+		return nil, diags
+	}
+	g.Subsume(&instG.AcyclicGraph.Graph)
+
+	return instanceAddrs, diags
+}
+
+func (n *nodeExpandActionDeclaration) actionInstanceSubgraph(addr addrs.AbsAction, instanceAddrs []addrs.AbsActionInstance) (*Graph, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+
+	// Start creating the steps
+	steps := []GraphTransformer{
```

Comment:
This whole pattern of creating an expansion graph with transformers and such is overkill for most things that aren't resources (and we should probably remove it for resources too). If it helps you can just add the nodes directly like how import nodes are done. Otherwise you can leave it and we can review this later, since we probably should simplify the resource expansion in the same way too.